### PR TITLE
Documentation: manipulating projections

### PR DESCRIPTION
**Pull request by morozovsk**
info about manipulating projections based on RFC #14730

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Documentation (changelog entry is not required)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

...

Detailed description / Documentation draft:

...

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: docs/en/sql-reference/statements/alter/projection.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+-   `ALTER TABLE [db].name DROP PROJECTION name` - Removes projection description from tables metadata and deletes projection files from disk.
+
+-   `ALTER TABLE [db.]table MATERIALIZE PROJECTION name IN PARTITION partition_name` - The query rebuilds the projection `name` in the partition `partition_name`. Implemented as a [mutation](../../../../sql-reference/statements/alter/index.md#mutations).
+
+-   `ALTER TABLE [db.]table CLEAR PROJECTION name IN PARTITION partition_name`
+
+The first two commands are lightweight in a sense that they only change metadata or remove files.
+
+Also, they are replicated, syncing indices metadata via ZooKeeper.
```
Comment (amosbird) [inline docs/en/sql-reference/statements/alter/projection.md:20]: indices -> projections

File: docs/en/sql-reference/statements/alter/projection.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+The following operations are available:
+
+-   `ALTER TABLE [db].name ADD PROJECTION name AS SELECT <COLUMN LIST EXPR> [WHERE] [GROUP BY] [ORDER BY]` - Adds projection description to tables metadata.
+
+-   `ALTER TABLE [db].name DROP PROJECTION name` - Removes projection description from tables metadata and deletes projection files from disk.
+
+-   `ALTER TABLE [db.]table MATERIALIZE PROJECTION name IN PARTITION partition_name` - The query rebuilds the projection `name` in the partition `partition_name`. Implemented as a [mutation](../../../../sql-reference/statements/alter/index.md#mutations).
+
+-   `ALTER TABLE [db.]table CLEAR PROJECTION name IN PARTITION partition_name`
```
Comment (amosbird) [inline docs/en/sql-reference/statements/alter/projection.md:16]: -   `ALTER TABLE [db.]table CLEAR PROJECTION name [IN PARTITION partition_name]` - Deletes projection files from disk without removing description.

File: docs/en/sql-reference/statements/alter/projection.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+The first two commands are lightweight in a sense that they only change metadata or remove files.
```
Comment (amosbird) [inline docs/en/sql-reference/statements/alter/projection.md:18]: `CLEAR` is also lightweight

**Review by morozovsk — COMMENTED** [review]
fixed. I also deleted WHERE because it is currently not supported. thanks for review and for projections.

**Comment by edwinkong** [comment]
In ClickHouse version 21.6.3.14 revision 54448 the correct syntax of adding projection should be 
`ALTER TABLE [db].name ADD PROJECTION name (SELECT <COLUMN LIST EXPR> [GROUP BY] [ORDER BY])`

Moreover, I think it maybe worth mentioning if one manually cleared a projection in certain partition of the table, that part of data would be missing from the result returned by the query which used that projection.
 
In other words, the feature mentioned in #14730, part Query Routing Process
> The query pipeline which uses projections will be different from the one that uses the original parts. if the projection is absent in some parts, we can add the pipeline to "project" it on the fly.

is not implemented in that version of ClickHouse yet.

**Comment by amosbird** [comment]
> Moreover, I think it maybe worth mentioning if one manually cleared a projection in certain partition of the table, that part of data would be missing from the result returned by the query which used that projection.

How is this feature not implemented? Could you give a minimal example that doesn't work as expected?

**Comment by edwinkong** [comment]
> > Moreover, I think it may be worth mentioning if one manually cleared a projection in certain partition of the table, that part of data would be missing from the result returned by the query which used that projection.
> 
> How is this feature not implemented? Could you give a minimal example that doesn't work as expected? @edwinkong


```
CREATE TABLE p (
    a UInt32,
    b UInt32,
    PROJECTION proj__sum_b (
        SELECT
            sum(b)
        GROUP BY intDiv(a, 100)
    )
) 
ENGINE MergeTree()
PARTITION BY intDiv(a, 100)
PRIMARY KEY a
ORDER BY a;

INSERT INTO p SELECT number, number FROM numbers(1000);

SELECT sum(b) FROM p GROUP BY intDiv(a, 100);
-- 10 rows in set
SELECT count(b) FROM p GROUP BY intDiv(a, 100);
-- 10 rows in set

ALTER TABLE p CLEAR PROJECTION proj__sum_b IN PARTITION 1;

SELECT sum(b) FROM p GROUP BY intDiv(a, 100);
-- 9 rows in set
SELECT count(b) FROM p GROUP BY intDiv(a, 100);
-- 10 rows in set
```

So if the projection is absent in some parts, those parts would be ignored by the query pipeline in ClickHouse version 21.6.3.14 revision 54448. I am sorry if I misunderstood that statement at the beginning...

**Comment by amosbird** [comment]
I cannot reproduce the result with current master @edwinkong

**Comment by edwinkong** [comment]
> I cannot reproduce the result with current master @edwinkong

Thank you for your information! I just tested on 21.7 and this issue is fixed. If you are interested, I included the query traces captured from the log of 21.6.3.14 and 21.7.2.7 below. I don't know if this should be included as a bugfix in the release note, anyway, thank you very much for adding this feature to ClickHouse!

Version 21.7.2.7
```
2021.07.12 10:16:46.802854 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> executeQuery: (from 127.0.0.1:45164, using production parser) SELECT sum(b) FROM p GROUP BY intDiv(a, 100);
2021.07.12 10:16:46.803495 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> ContextAccess (default): Access granted: SELECT(a, b) ON default.p
2021.07.12 10:16:46.803727 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> ContextAccess (default): Access granted: SELECT(a, b) ON default.p
2021.07.12 10:16:46.803790 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> default.p (947be7a4-2716-49c5-947b-e7a4271659c5) (SelectExecutor): Choose aggregate projection proj__sum_b
2021.07.12 10:16:46.803817 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> default.p (947be7a4-2716-49c5-947b-e7a4271659c5) (SelectExecutor): projection required columns: intDiv(a, 100), sum(b)
2021.07.12 10:16:46.803995 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> default.p (947be7a4-2716-49c5-947b-e7a4271659c5) (SelectExecutor): Key condition: unknown
2021.07.12 10:16:46.804203 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> default.p (947be7a4-2716-49c5-947b-e7a4271659c5) (SelectExecutor): MinMax index condition: unknown
2021.07.12 10:16:46.804887 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> default.p (947be7a4-2716-49c5-947b-e7a4271659c5) (SelectExecutor): Selected 9/9 parts by partition key, 9 parts by primary key, 9/9 marks by primary key, 9 marks to read from 9 ranges
2021.07.12 10:16:46.805028 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> default.p (947be7a4-2716-49c5-947b-e7a4271659c5) (SelectExecutor): Reading approx. 9 rows with 9 streams
2021.07.12 10:16:46.805740 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> default.p (947be7a4-2716-49c5-947b-e7a4271659c5) (SelectExecutor): Key condition: unknown
2021.07.12 10:16:46.805863 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> default.p (947be7a4-2716-49c5-947b-e7a4271659c5) (SelectExecutor): MinMax index condition: unknown
2021.07.12 10:16:46.805906 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> default.p (947be7a4-2716-49c5-947b-e7a4271659c5) (SelectExecutor): Selected 1/1 parts by partition key, 1 parts by primary key, 1/1 marks by primary key, 1 marks to read from 1 ranges
2021.07.12 10:16:46.805954 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> MergeTreeSelectProcessor: Reading 1 ranges from part 1_2_2_0_11, approx. 100 rows starting from 0
2021.07.12 10:16:46.806007 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> InterpreterSelectQuery: FetchColumns -> WithMergeableState
2021.07.12 10:16:46.806316 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> InterpreterSelectQuery: WithMergeableState -> Complete
2021.07.12 10:16:46.807489 [ 187 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> AggregatingTransform: Aggregating
2021.07.12 10:16:46.807533 [ 187 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> Aggregator: Aggregation method: key32
2021.07.12 10:16:46.807575 [ 99 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> AggregatingTransform: Aggregating
2021.07.12 10:16:46.807595 [ 193 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> AggregatingTransform: Aggregating
2021.07.12 10:16:46.807584 [ 195 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> AggregatingTransform: Aggregating
2021.07.12 10:16:46.807639 [ 193 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> Aggregator: Aggregation method: key32
2021.07.12 10:16:46.807663 [ 195 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> Aggregator: Aggregation method: key32
2021.07.12 10:16:46.807673 [ 222 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> AggregatingTransform: Aggregating
2021.07.12 10:16:46.807722 [ 222 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> Aggregator: Aggregation method: key32
2021.07.12 10:16:46.807800 [ 222 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> AggregatingTransform: Aggregated. 100 to 1 rows (from 800.00 B) in 0.001507223 sec. (66347.183 rows/sec., 518.34 KiB/sec.)
2021.07.12 10:16:46.807810 [ 187 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> AggregatingTransform: Aggregated. 2 to 2 rows (from 8.02 KiB) in 0.00164682 sec. (1214.462 rows/sec., 4.76 MiB/sec.)
2021.07.12 10:16:46.807805 [ 194 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> AggregatingTransform: Aggregated. 0 to 0 rows (from 0.00 B) in 0.001567755 sec. (0.000 rows/sec., 0.00 B/sec.)
2021.07.12 10:16:46.807865 [ 222 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> AggregatingTransform: Aggregated. 0 to 0 rows (from 0.00 B) in 0.001673106 sec. (0.000 rows/sec., 0.00 B/sec.)
2021.07.12 10:16:46.807641 [ 219 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> AggregatingTransform: Aggregating
2021.07.12 10:16:46.807917 [ 219 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> Aggregator: Aggregation method: key32
2021.07.12 10:16:46.807924 [ 195 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> AggregatingTransform: Aggregated. 2 to 2 rows (from 8.02 KiB) in 0.00171881 sec. (1163.596 rows/sec., 4.56 MiB/sec.)
2021.07.12 10:16:46.807629 [ 99 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> Aggregator: Aggregation method: key32
2021.07.12 10:16:46.807882 [ 187 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> AggregatingTransform: Aggregated. 0 to 0 rows (from 0.00 B) in 0.001657525 sec. (0.000 rows/sec., 0.00 B/sec.)
2021.07.12 10:16:46.807983 [ 219 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> AggregatingTransform: Aggregated. 1 to 1 rows (from 4.01 KiB) in 0.001716651 sec. (582.530 rows/sec., 2.28 MiB/sec.)
2021.07.12 10:16:46.807993 [ 193 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> AggregatingTransform: Aggregated. 2 to 2 rows (from 8.02 KiB) in 0.001807072 sec. (1106.763 rows/sec., 4.34 MiB/sec.)
2021.07.12 10:16:46.808001 [ 99 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> AggregatingTransform: Aggregated. 1 to 1 rows (from 4.01 KiB) in 0.001785782 sec. (559.979 rows/sec., 2.19 MiB/sec.)
2021.07.12 10:16:46.807933 [ 223 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> AggregatingTransform: Aggregating
2021.07.12 10:16:46.808084 [ 223 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> Aggregator: Aggregation method: key32
2021.07.12 10:16:46.808137 [ 223 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> AggregatingTransform: Aggregated. 1 to 1 rows (from 4.01 KiB) in 0.00186552 sec. (536.044 rows/sec., 2.10 MiB/sec.)
2021.07.12 10:16:46.808172 [ 223 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Trace> Aggregator: Merging aggregated data
2021.07.12 10:16:46.808997 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Information> executeQuery: Read 109 rows, 36.89 KiB in 0.006073531 sec., 17946 rows/sec., 5.93 MiB/sec.
2021.07.12 10:16:46.809037 [ 51 ] {436089c7-8bca-4f82-9b0d-2c4f61986dba} <Debug> MemoryTracker: Peak memory usage (for query): 0.00 B.
```

Version 21.6.3.14
```
2021.07.12 10:00:33.275445 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> executeQuery: (from 127.0.0.1:54124, using production parser) SELECT sum(b) FROM p GROUP BY intDiv(a, 100);
2021.07.12 10:00:33.276564 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Trace> ContextAccess (default): Access granted: SELECT(a, b) ON default.p
2021.07.12 10:00:33.276775 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> default.p (7351bc70-775b-4a58-b351-bc70775b2a58) (SelectExecutor): Key condition: unknown
2021.07.12 10:00:33.276983 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> default.p (7351bc70-775b-4a58-b351-bc70775b2a58) (SelectExecutor): MinMax index condition: unknown
2021.07.12 10:00:33.281912 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> default.p (7351bc70-775b-4a58-b351-bc70775b2a58) (SelectExecutor): Selected 10/10 parts by partition key, 10 parts by primary key, 10/10 marks by primary key, 10 marks to read from 10 ranges
2021.07.12 10:00:33.282227 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> default.p (7351bc70-775b-4a58-b351-bc70775b2a58) (SelectExecutor): Key condition: unknown
2021.07.12 10:00:33.282418 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> default.p (7351bc70-775b-4a58-b351-bc70775b2a58) (SelectExecutor): MinMax index condition: unknown
2021.07.12 10:00:33.287102 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> default.p (7351bc70-775b-4a58-b351-bc70775b2a58) (SelectExecutor): Selected 9/9 parts by partition key, 9 parts by primary key, 9/9 marks by primary key, 9 marks to read from 9 ranges
2021.07.12 10:00:33.287393 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> default.p (7351bc70-775b-4a58-b351-bc70775b2a58) (SelectExecutor): Key condition: unknown
2021.07.12 10:00:33.287576 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> default.p (7351bc70-775b-4a58-b351-bc70775b2a58) (SelectExecutor): MinMax index condition: unknown
2021.07.12 10:00:33.287632 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> default.p (7351bc70-775b-4a58-b351-bc70775b2a58) (SelectExecutor): Selected 1/1 parts by partition key, 1 parts by primary key, 1/1 marks by primary key, 1 marks to read from 1 ranges
2021.07.12 10:00:33.287781 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Trace> ContextAccess (default): Access granted: SELECT(a, b) ON default.p
2021.07.12 10:00:33.287892 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> default.p (7351bc70-775b-4a58-b351-bc70775b2a58) (SelectExecutor): Choose aggregate projection proj__sum_b
2021.07.12 10:00:33.287927 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> default.p (7351bc70-775b-4a58-b351-bc70775b2a58) (SelectExecutor): projection required columns: intDiv(a, 100), sum(b)
2021.07.12 10:00:33.288252 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> default.p (7351bc70-775b-4a58-b351-bc70775b2a58) (SelectExecutor): Reading approx. 9 rows with 9 streams
2021.07.12 10:00:33.289167 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Trace> InterpreterSelectQuery: FetchColumns -> WithMergeableState
2021.07.12 10:00:33.289685 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Trace> InterpreterSelectQuery: WithMergeableState -> Complete
2021.07.12 10:00:33.294468 [ 96 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Trace> AggregatingTransform: Aggregating
2021.07.12 10:00:33.294537 [ 96 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Trace> Aggregator: Aggregation method: key32
2021.07.12 10:00:33.295199 [ 54 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Trace> AggregatingTransform: Aggregating
2021.07.12 10:00:33.295232 [ 92 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Trace> AggregatingTransform: Aggregating
2021.07.12 10:00:33.295284 [ 92 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Trace> Aggregator: Aggregation method: key32
2021.07.12 10:00:33.295362 [ 110 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> AggregatingTransform: Aggregated. 0 to 0 rows (from 0.00 B) in 0.005722228 sec. (0.000 rows/sec., 0.00 B/sec.)
2021.07.12 10:00:33.295614 [ 92 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> AggregatingTransform: Aggregated. 2 to 2 rows (from 8.02 KiB) in 0.006034962 sec. (331.402 rows/sec., 1.30 MiB/sec.)
2021.07.12 10:00:33.295690 [ 92 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> AggregatingTransform: Aggregated. 0 to 0 rows (from 0.00 B) in 0.006094527 sec. (0.000 rows/sec., 0.00 B/sec.)
2021.07.12 10:00:33.295710 [ 96 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> AggregatingTransform: Aggregated. 5 to 5 rows (from 20.06 KiB) in 0.006251467 sec. (799.812 rows/sec., 3.13 MiB/sec.)
2021.07.12 10:00:33.295738 [ 58 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> AggregatingTransform: Aggregated. 0 to 0 rows (from 0.00 B) in 0.006215628 sec. (0.000 rows/sec., 0.00 B/sec.)
2021.07.12 10:00:33.295747 [ 92 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> AggregatingTransform: Aggregated. 0 to 0 rows (from 0.00 B) in 0.006208511 sec. (0.000 rows/sec., 0.00 B/sec.)
2021.07.12 10:00:33.295616 [ 110 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Trace> AggregatingTransform: Aggregating
2021.07.12 10:00:33.295780 [ 96 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> AggregatingTransform: Aggregated. 0 to 0 rows (from 0.00 B) in 0.006222767 sec. (0.000 rows/sec., 0.00 B/sec.)
2021.07.12 10:00:33.295802 [ 110 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Trace> Aggregator: Aggregation method: key32
2021.07.12 10:00:33.295855 [ 110 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> AggregatingTransform: Aggregated. 1 to 1 rows (from 4.01 KiB) in 0.006354605 sec. (157.366 rows/sec., 631.31 KiB/sec.)
2021.07.12 10:00:33.295238 [ 54 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Trace> Aggregator: Aggregation method: key32
2021.07.12 10:00:33.295920 [ 54 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> AggregatingTransform: Aggregated. 1 to 1 rows (from 4.01 KiB) in 0.006451215 sec. (155.010 rows/sec., 621.85 KiB/sec.)
2021.07.12 10:00:33.296898 [ 111 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> AggregatingTransform: Aggregated. 0 to 0 rows (from 0.00 B) in 0.007388546 sec. (0.000 rows/sec., 0.00 B/sec.)
2021.07.12 10:00:33.296932 [ 111 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Trace> Aggregator: Merging aggregated data
2021.07.12 10:00:33.302104 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Information> executeQuery: Read 9 rows, 36.11 KiB in 0.026559797 sec., 338 rows/sec., 1.33 MiB/sec.
2021.07.12 10:00:33.302167 [ 51 ] {60f51d4c-7fad-4c03-a26c-4d2c6ec79b1e} <Debug> MemoryTracker: Peak memory usage (for query): 0.00 B.
```

**Comment by amosbird** [comment]
@edwinkong The related fix is https://github.com/ClickHouse/ClickHouse/pull/25055

> anyway, thank you very much for adding this feature to ClickHouse!

My pleasure!

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
```
WARNING:mkdocs.structure.pages:Documentation file 'sql-reference/statements/alter/projection.md' contains a link to '../sql-reference/statements/alter/index.md' which is not found in the documentation files.
WARNING:mkdocs.structure.pages:Documentation file 'sql-reference/statements/alter/projection.md' contains a link to '../engines/table-engines/mergetree-family/mergetree.md' which is not found in the documentation files.
WARNING:mkdocs.structure.pages:Documentation file 'sql-reference/statements/alter/projection.md' contains a link to '../engines/table-engines/mergetree-family/replication.md' which is not found in the documentation files.
```

**Comment by olgarev** [comment]
Internal documentation ticket DOCSUP-12894
